File: engine_py/bazi.py
"""
Bazi (Four Pillars of Destiny) Calculator.
Rule-based cycle math — no AI.
"""

from datetime import datetime

STEMS = ["甲", "乙", "丙", "丁", "戊", "己", "庚", "辛", "壬", "癸"]
BRANCHES = ["子", "丑", "寅", "卯", "辰", "巳", "午", "未", "申", "酉", "戌", "亥"]

STEM_ELEMENT = {
    "甲": "Wood", "乙": "Wood", "丙": "Fire", "丁": "Fire",
    "戊": "Earth", "己": "Earth", "庚": "Metal", "辛": "Metal",
    "壬": "Water", "癸": "Water",
}
BRANCH_ELEMENT = {
    "子": "Water", "丑": "Earth", "寅": "Wood", "卯": "Wood",
    "辰": "Earth", "巳": "Fire", "午": "Fire", "未": "Earth",
    "申": "Metal", "酉": "Metal", "戌": "Earth", "亥": "Water",
}
BRANCH_ANIMAL = {
    "子": "Rat", "丑": "Ox", "寅": "Tiger", "卯": "Rabbit",
    "辰": "Dragon", "巳": "Snake", "午": "Horse", "未": "Goat",
    "申": "Monkey", "酉": "Rooster", "戌": "Dog", "亥": "Pig",
}

# Approximate solar term boundary dates (month -> (month_num, day))
SOLAR_TERMS = [
    (2, 4), (3, 6), (4, 5), (5, 6), (6, 6), (7, 7),
    (8, 8), (9, 8), (10, 8), (11, 7), (12, 7), (1, 6),
]
# Month branch for each solar term
MONTH_BRANCHES = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]  # 寅=0 for 立春, etc.

GENERATES = {"Wood": "Fire", "Fire": "Earth", "Earth": "Metal",
             "Metal": "Water", "Water": "Wood"}
CONTROLS = {"Wood": "Earth", "Earth": "Water", "Water": "Fire",
            "Fire": "Metal", "Metal": "Wood"}


def julian_day(y, m, d):
    """Julian Day Number."""
    if m <= 2:
        y -= 1
        m += 12
    A = y // 100
    B = 2 - A + A // 4
    return int(365.25 * (y + 4716)) + int(30.6001 * (m + 1)) + d + B - 1524


def calculate_bazi(birth_date, birth_time):
    """Calculate Four Pillars of Destiny.

    Args:
        birth_date: "YYYY-MM-DD"
        birth_time: "HH:MM" (24h)
    """
    y, m, d = map(int, birth_date.split("-"))
    h, mi = map(int, birth_time.split(":"))
    hour = h + mi / 60

    # Year pillar
    year_stem = (y - 4) % 10
    year_branch = (y - 4) % 12

    # Month branch from solar term
    # Solar terms cycle: Feb4 -> 寅, Mar6 -> 卯, ..., Jan6 -> 丑
    # Find the most recent term ON or BEFORE the birth date
    month_branch_idx = 0  # default: 立春
    # Check terms in chronological order (Feb4, Mar6, ... Jan6)
    # Terms before Oct are months 2-8; terms after Sep wrap to Jan
    # We need to find the last term whose date is <= birth date
    for i in range(len(SOLAR_TERMS)):
        sm, sd = SOLAR_TERMS[i]
        # Handle wrap: Jan6 (index 11) is AFTER Dec7 (index 10)
        # So Jan6's month=1 but it's actually the last term of the cycle
        if i == 11:  # 小寒 (Jan 6) — special case
            if m > 12 or (m == 1 and d >= 6):
                month_branch_idx = 11
        elif i < 11:  # Regular months Feb-Dec
            if m > sm or (m == sm and d >= sd):
                month_branch_idx = i
    # Convert solar term index to branch index: 寅=BRANCHES[2]
    term_to_branch = [(i + 2) % 12 for i in range(12)]
    month_branch = term_to_branch[month_branch_idx]

    month_stem = ((year_stem) * 2 + month_branch_idx) % 10

    # Day pillar from JD
    jd = julian_day(y, m, d)
    day_stem = (int(jd + 0.5) + 9) % 10
    day_branch = (int(jd + 0.5) + 1) % 12

    # Hour pillar
    if hour >= 23 or hour < 1:
        hour_branch = 0
    elif hour < 3:
        hour_branch = 1
    elif hour < 5:
        hour_branch = 2
    elif hour < 7:
        hour_branch = 3
    elif hour < 9:
        hour_branch = 4
    elif hour < 11:
        hour_branch = 5
    elif hour < 13:
        hour_branch = 6
    elif hour < 15:
        hour_branch = 7
    elif hour < 17:
        hour_branch = 8
    elif hour < 19:
        hour_branch = 9
    elif hour < 21:
        hour_branch = 10
    else:
        hour_branch = 11

    hour_stem = (day_stem * 2 + hour_branch) % 10

    # Day master
    day_master = STEMS[day_stem]

    # Compose pillars
    pillars = {
        "year": {"stem": STEMS[year_stem], "branch": BRANCHES[year_branch],
                 "element": STEM_ELEMENT[STEMS[year_stem]]},
        "month": {"stem": STEMS[month_stem], "branch": BRANCHES[month_branch],
                  "element": STEM_ELEMENT[STEMS[month_stem]]},
        "day": {"stem": STEMS[day_stem], "branch": BRANCHES[day_branch],
                "element": STEM_ELEMENT[STEMS[day_stem]]},
        "hour": {"stem": STEMS[hour_stem], "branch": BRANCHES[hour_branch],
                 "element": STEM_ELEMENT[STEMS[hour_stem]]},
    }

    # Element analysis
    dm_elem = STEM_ELEMENT[day_master]
    all_elements = []
    for p in ["year", "month", "day", "hour"]:
        all_elements.append(STEM_ELEMENT[pillars[p]["stem"]])
        all_elements.append(BRANCH_ELEMENT[pillars[p]["branch"]])

    # Favorable = element that generates DM + DM itself
    favorable = [dm_elem]
    gen = GENERATES.get(dm_elem)
    if gen and gen not in favorable:
        favorable.insert(0, gen)

    # Unfavorable = element that controls DM + element DM controls
    unfavorable = []
    ctrl = CONTROLS.get(dm_elem)
    if ctrl:
        unfavorable.append(ctrl)
    ctrl_by = next((k for k, v in CONTROLS.items() if v == dm_elem), None)
    if ctrl_by and ctrl_by not in unfavorable:
        unfavorable.append(ctrl_by)

    # Luck pillar (10-year cycles from start age)
    now = datetime.now()
    age = now.year - y - (1 if (now.month, now.day) < (m, d) else 0)
    # Approximate: start age based on gender  (simplified)
    start_age = 3  # common default
    luck_idx = max(0, (age - start_age) // 10) if age >= start_age else 0

    luck_stem = (month_stem + luck_idx + 1) % 10
    luck_branch = (month_branch + luck_idx + 1) % 12

    luck_pillar = f"{STEMS[luck_stem]}{BRANCHES[luck_branch]}"
    luck_elem = STEM_ELEMENT[STEMS[luck_stem]]

    # Element relationship
    if luck_elem == dm_elem:
        rel = f"{luck_elem} (same as Day Master) — reinforcing"
    elif GENERATES.get(luck_elem) == dm_elem:
        rel = f"{luck_elem} (generates Day Master) — supportive"
    elif CONTROLS.get(luck_elem) == dm_elem:
        rel = f"{luck_elem} (controls Day Master) — challenging"
    else:
        rel = f"{luck_elem} — neutral"

    luck_start = start_age + luck_idx * 10

    return {
        "four_pillars": pillars,
        "day_master": day_master,
        "favorable_elements": favorable,
        "unfavorable_elements": unfavorable,
        "current_luck_pillar": {
            "pillar": luck_pillar,
            "age_range": f"{luck_start}-{luck_start + 9}",
            "element_relationship": rel,
        },
    }

File: src/lib/astro/bazi.ts
// ─── Bazi (Four Pillars) Calculator ────────────────────────────────
// Sexagenary cycle calculations from birth date/time.
// 100% accurate — deterministic cycle math.

export interface Pillar {
  stem: string;
  branch: string;
  element: string;
}

export interface BaziResult {
  four_pillars: {
    year: Pillar;
    month: Pillar;
    day: Pillar;
    hour: Pillar;
  };
  day_master: string;
  favorable_elements: string[];
  unfavorable_elements: string[];
  current_luck_pillar: {
    pillar: string;
    age_range: string;
    element_relationship: string;
  };
}

const STEMS = ["甲", "乙", "丙", "丁", "戊", "己", "庚", "辛", "壬", "癸"];
const BRANCHES = ["子", "丑", "寅", "卯", "辰", "巳", "午", "未", "申", "酉", "戌", "亥"];

const STEM_ELEMENTS: Record<string, string> = {
  "甲": "Wood", "乙": "Wood",
  "丙": "Fire", "丁": "Fire",
  "戊": "Earth", "己": "Earth",
  "庚": "Metal", "辛": "Metal",
  "壬": "Water", "癸": "Water",
};

const BRANCH_ELEMENTS: Record<string, string> = {
  "子": "Water", "丑": "Earth", "寅": "Wood", "卯": "Wood",
  "辰": "Earth", "巳": "Fire", "午": "Fire", "未": "Earth",
  "申": "Metal", "酉": "Metal", "戌": "Earth", "亥": "Water",
};

const STEM_YIN_YANG: Record<string, string> = {
  "甲": "Yang", "乙": "Yin", "丙": "Yang", "丁": "Yin",
  "戊": "Yang", "己": "Yin", "庚": "Yang", "辛": "Yin",
  "壬": "Yang", "癸": "Yin",
};

// Element cycles for determining favorability
const GENERATES: Record<string, string> = {
  "Wood": "Fire", "Fire": "Earth", "Earth": "Metal",
  "Metal": "Water", "Water": "Wood",
};
const CONTROLS: Record<string, string> = {
  "Wood": "Earth", "Earth": "Water", "Water": "Fire",
  "Fire": "Metal", "Metal": "Wood",
};

// Approximate solar term dates (month boundaries for Bazi)
// Each entry: [month (Jan=1), day] → branch and stem offset
const SOLAR_TERMS: { month: number; day: number; branch: number; stemOffset: number }[] = [
  { month: 2, day: 4, branch: 0, stemOffset: 0 },   // 立春 → 寅
  { month: 3, day: 6, branch: 1, stemOffset: 2 },    // 惊蛰 → 卯
  { month: 4, day: 5, branch: 2, stemOffset: 4 },    // 清明 → 辰
  { month: 5, day: 6, branch: 3, stemOffset: 6 },    // 立夏 → 巳
  { month: 6, day: 6, branch: 4, stemOffset: 8 },    // 芒种 → 午
  { month: 7, day: 7, branch: 5, stemOffset: 0 },    // 小暑 → 未 (reset stem offset)
  { month: 8, day: 8, branch: 6, stemOffset: 2 },    // 立秋 → 申
  { month: 9, day: 8, branch: 7, stemOffset: 4 },    // 白露 → 酉
  { month: 10, day: 8, branch: 8, stemOffset: 6 },   // 寒露 → 戌
  { month: 11, day: 7, branch: 9, stemOffset: 8 },   // 立冬 → 亥
  { month: 12, day: 7, branch: 10, stemOffset: 0 },  // 大雪 → 子
  { month: 1, day: 6, branch: 11, stemOffset: 2 },   // 小寒 → 丑
];

function julianDay(year: number, month: number, day: number): number {
  let y = year, m = month;
  if (m <= 2) { y--; m += 12; }
  const A = Math.floor(y / 100);
  const B = 2 - A + Math.floor(A / 4);
  return Math.floor(365.25 * (y + 4716)) + Math.floor(30.6001 * (m + 1)) + day + B - 1524.5;
}

export function calculateBazi(birthDate: string, birthTime: string): BaziResult {
  // Parse date and time
  const [y, mon, d] = birthDate.split("-").map(Number);
  const [h, min] = birthTime.split(":").map(Number);
  const hour = h + min / 60;

  // Year pillar: sexagenary cycle
  // Year 4 CE = 甲子 (stem=0, branch=0)
  const yearStem = (y - 4) % 10;
  const yearBranch = (y - 4) % 12;
  const yearStemChar = STEMS[(yearStem + 10) % 10];
  const yearBranchChar = BRANCHES[(yearBranch + 12) % 12];

  // Month pillar: based on solar terms
  let monthBranch = 0;
  let monthStemOffset = 0;
  for (let i = SOLAR_TERMS.length - 1; i >= 0; i--) {
    const st = SOLAR_TERMS[i];
    if (mon > st.month || (mon === st.month && d >= st.day)) {
      monthBranch = st.branch;
      monthStemOffset = st.stemOffset;
      break;
    }
  }
  const monthStemIdx = ((yearStem + 10) % 10) * 2 + monthStemOffset;
  const monthStemChar = STEMS[monthStemIdx % 10];
  const monthBranchChar = BRANCHES[monthBranch];

  // Day pillar: from Julian day
  const jd = julianDay(y, mon, d);
  const dayStem = ((Math.floor(jd + 0.5) + 9) % 10 + 10) % 10;
  const dayBranch = ((Math.floor(jd + 0.5) + 1) % 12 + 12) % 12;
  const dayStemChar = STEMS[dayStem % 10];
  const dayBranchChar = BRANCHES[dayBranch % 12];

  // Hour pillar: Chinese double-hour (2-hour blocks)
  // 子 = 23:00-00:59, 丑 = 01:00-02:59, ..., 亥 = 21:00-22:59
  let hourBranch: number;
  if (hour >= 23 || hour < 1) hourBranch = 0;
  else if (hour < 3) hourBranch = 1;
  else if (hour < 5) hourBranch = 2;
  else if (hour < 7) hourBranch = 3;
  else if (hour < 9) hourBranch = 4;
  else if (hour < 11) hourBranch = 5;
  else if (hour < 13) hourBranch = 6;
  else if (hour < 15) hourBranch = 7;
  else if (hour < 17) hourBranch = 8;
  else if (hour < 19) hourBranch = 9;
  else if (hour < 21) hourBranch = 10;
  else hourBranch = 11;

  const hourStemIdx = (dayStem * 2 + hourBranch) % 10;
  const hourStemChar = STEMS[hourStemIdx];
  const hourBranchChar = BRANCHES[hourBranch];

  // Day master
  const dayMaster = dayStemChar;

  // Element analysis
  const dmElement = STEM_ELEMENTS[dayMaster] || "Wood";
  const allStems = [yearStemChar, monthStemChar, dayStemChar, hourStemChar];
  const allBranches = [yearBranchChar, monthBranchChar, dayBranchChar, hourBranchChar];

  // Count elements
  const elementCount: Record<string, number> = { Wood: 0, Fire: 0, Earth: 0, Metal: 0, Water: 0 };
  for (const s of allStems) elementCount[STEM_ELEMENTS[s]]++;
  for (const b of allBranches) elementCount[BRANCH_ELEMENTS[b]]++;

  // Favorable = elements that generate or are same as DM
  // Unfavorable = elements that control or are controlled by DM
  const favorable: string[] = [dmElement];
  const unfavorable: string[] = [];

  const generator = Object.entries(GENERATES).find(([_, v]) => v === dmElement)?.[0];
  if (generator) favorable.push(generator);

  const controller = CONTROLS[dmElement];
  if (controller && !unfavorable.includes(controller)) unfavorable.push(controller);

  const controlledBy = Object.entries(CONTROLS).find(([_, v]) => v === dmElement)?.[0];
  if (controlledBy && !unfavorable.includes(controlledBy)) unfavorable.push(controlledBy);

  // Current luck pillar (10-year cycles starting from birth)
  // Simplified: luck pillar starts at age based on gender and year stem yin/yang
  // Using a common approximation: start age = 1-10 depending on birth
  const dmYinYang = STEM_YIN_YANG[dayMaster];
  const yearYinYang = STEM_YIN_YANG[yearStemChar];
  const forward = (dmYinYang === "Yang" && yearYinYang === "Yang") || (dmYinYang === "Yin" && yearYinYang === "Yin");
  const startAge = forward ? (yearBranch + 1) % 10 + 1 : 10 - ((yearBranch) % 10);

  // Current age → current luck pillar
  const now = new Date();
  const age = now.getFullYear() - y - (now.getMonth() + 1 < mon || (now.getMonth() + 1 === mon && now.getDate() < d) ? 1 : 0);
  const luckIndex = Math.floor((age - startAge) / 10);
  const luckStartAge = startAge + luckIndex * 10;

  const luckStem = forward
    ? (monthStemIdx + luckIndex + 1) % 10
    : (monthStemIdx - luckIndex - 1 + 20) % 10;
  const luckBranch = forward
    ? (monthBranch + luckIndex + 1) % 12
    : (monthBranch - luckIndex - 1 + 24) % 12;

  const luckPillar = `${STEMS[luckStem]}${BRANCHES[luckBranch]}`;
  const luckElement = STEM_ELEMENTS[STEMS[luckStem]];

  let elementRel = "neutral";
  if (GENERATES[luckElement] === dmElement) elementRel = "supportive (generates Day Master)";
  else if (CONTROLS[luckElement] === dmElement) elementRel = "challenging (controls Day Master)";
  else if (luckElement === dmElement) elementRel = "reinforcing (same as Day Master)";
  else if (GENERATES[dmElement] === luckElement) elementRel = "draining (Day Master generates it)";
  else if (CONTROLS[dmElement] === luckElement) elementRel = "restrictive (Day Master controls it)";

  return {
    four_pillars: {
      year: { stem: yearStemChar, branch: yearBranchChar, element: STEM_ELEMENTS[yearStemChar] },
      month: { stem: monthStemChar, branch: monthBranchChar, element: STEM_ELEMENTS[monthStemChar] },
      day: { stem: dayStemChar, branch: dayBranchChar, element: STEM_ELEMENTS[dayStemChar] },
      hour: { stem: hourStemChar, branch: hourBranchChar, element: STEM_ELEMENTS[hourStemChar] },
    },
    day_master: dayMaster,
    favorable_elements: favorable,
    unfavorable_elements: unfavorable,
    current_luck_pillar: {
      pillar: luckPillar,
      age_range: `${luckStartAge}-${luckStartAge + 9}`,
      element_relationship: `${luckElement} (${STEMS[luckStem]}) — ${elementRel}`,
    },
  };
}
